### PR TITLE
feature/APP-785_aragon-abstract-plugin

### PR DESCRIPTION
**Pull request by novaknole**
This brings abstract aragon plugin that should be inheritted by the plugin developer's plugins.

This currently contains `AragonUpgradablePlugin, AragonPlugin`.

1.  dev deploys a plugin with our PluginERC1967Proxy - This includes `dao` slot feature, so `auth` is available. (dev inherits from `AragonPluginUpgradable`)
2. dev deploys a plugin with our PluginERC1967Proxy - This includes `dao` slot feature, so `auth` is available. (dev inherits from `AragonPlugin`) but this doesn't include UUPSUpgradable !
3. dev deploys a plugin directly with no proxy and we might still want to support a feature where he inherits from `AragonPlugin`, but since it's deployed without proxy, dao doesn't exist on a slot. (Remaining)

### CODE REVIEW

File: packages/contracts/contracts/utils/AppStorage.sol [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+pragma solidity 0.8.10;
+
+import "./UnstructuredStorage.sol";
+import "../core/IDAO.sol";
+
+contract AppStorage {
+    using UnstructuredStorage for bytes32;
+    
+    /* Hardcoded constants to save gas
+    bytes32 internal constant DAO_POSITION = keccak256("aragonCore.appStorage.dao");
```
Comment (brickpop) [inline packages/contracts/contracts/utils/AppStorage.sol:12]: How about `core.storage.dao`? It's totally cosmetic, I know. But clean code starts here too :)

File: packages/contracts/contracts/utils/AppStorage.sol [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    */
+    bytes32 internal constant DAO_POSITION = 0xcc4b1e6cefe3fef82eb08a6a678aa17d9aaa5a1bf67e55309e9d2b3bbd029b0c;
+
+    function dao() public view returns (IDAO) {
+        return IDAO(DAO_POSITION.getStorageAddress());
+    }
+
+    function setDAO(address _dao) internal {
+        DAO_POSITION.setStorageAddress(_dao);
+    }
```
Comment (brickpop) [inline packages/contracts/contracts/utils/AppStorage.sol:22]: What are we exactly getting and setting? Natspec here sill be very useful

File: packages/contracts/contracts/utils/UUPSProxy.sol [hunk 3]
```diff
@@ -3,15 +3,19 @@
 pragma solidity 0.8.10;
 
 import "@openzeppelin/contracts/proxy/ERC1967/ERC1967Proxy.sol";
+import "./AppStorage.sol";
 
 /// @title UUPSProxy
-/// @author Aragon Association - 2022
+/// @author Giorgi Lagidze - Aragon Association - 2022
```
Comment (brickpop) [inline packages/contracts/contracts/utils/UUPSProxy.sol:9]: Aren't contributions recorded on Git? If the rest of files are authored by the association, I suggest we leave this one as is

File: packages/contracts/contracts/utils/UUPSProxy.sol [hunk 4]
```diff
@@ -3,15 +3,19 @@
...
 /// @notice This contract is only meant to be used for deploy purposes.
-/// TODO: Need to be fully tested, and decide to use the second address param as admin address.
-contract UUPSProxy is ERC1967Proxy {
+contract UUPSProxy is ERC1967Proxy, AppStorage {
```
Comment (brickpop) [inline packages/contracts/contracts/utils/UUPSProxy.sol:11]: Adding Natspec would also be helpful, even if it is for auditors

File: packages/contracts/contracts/core/plugin/AragonUpgradablePlugin.sol [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    } 
+
+    /// @notice Used by UI to detect what arguments it needs to provide to the update function.
+    /// @dev Expects append-only ABI for each update. 
+    /// If 1.x version needed `uint param1`, and 2.x only needed `address param2`
+    /// It's required 3.x have both in the ABI as appended in the same sequence = (uint param1, address param2)
+    /// @return the Human-Readable ABI for the update function.
+    function updateSignatureABI() external view virtual returns(string memory) { 
+        return "";
+    }
```
[the remark this answers is not in the document]
Comment (novaknole) [reply packages/contracts/contracts/core/plugin/AragonUpgradablePlugin.sol:80]: I don't want to make it `required`.  In the very first version, they don't have to define this function in their contract ! if we leave it empty, then they anyways have to define it which makes no sense..

File: packages/contracts/contracts/core/plugin/AragonUpgradablePlugin.sol [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+        uint8[3] calldata newVersion, 
+        bytes memory data
+    ) internal virtual;
+    
+    /// @notice allows to call `update` through `reinitializer` modifier if it was not called with this version yet.
+    /// @dev dev can override this and increment it if he/she wants update to be allowed to be called.
+    /// Used also by UI to detect when it can call `update` and when not.
+    /// ex: If the updateVersion(in the new base) > updateVersion(current one), update must be called.
+    /// @return the update version
+    function updateVersion() public virtual returns (uint8) { return 1; }
```
[the remark this answers is not in the document]
Comment (novaknole) [reply packages/contracts/contracts/core/plugin/AragonUpgradablePlugin.sol:55]: This `updateVersion`  doesn't relate exactly to when the updates happen. 

It's possible that contract code has changed, but `updateVersion` still stays whatever number was before. `updateVersion` only gets increased when there's a need to call `update initialization` function and It's super highly unlikely that this would happen more than 256 times..

File: packages/contracts/contracts/core/plugin/AragonUpgradablePlugin.sol [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+    /// @return the Human-Readable ABI for the update function.
+    function updateSignatureABI() external view virtual returns(string memory) { 
+        return "";
+    }
+
+    /// @notice Used by AragonUpgradablePlugin to reserve storage space in 
+    /// case of state variable additions for this contract.
+    /// @dev Note that after adding one or multiple state variables, 
+    /// _gap size below + all state variables this contract uses should be 50.
+    uint256[50] private __gap;
```
[the remark this answers is not in the document]
Comment (novaknole) [reply packages/contracts/contracts/core/plugin/AragonUpgradablePlugin.sol:86]: Nope. if you do it, when the next update comes in and you want to add a variable, where do you put that variable ? below `_gap` or on top of `_gap`?  

This just eliminates the thought process more about storage slots from us. OZ puts it the same way which I completely agree.

File: packages/contracts/contracts/utils/AppStorage.sol [hunk 8]
```diff
@@ -0,0 +1,24 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity 0.8.10;
+
+import "./UnstructuredStorage.sol";
+import "../core/IDAO.sol";
+
+contract AppStorage {
```
[the remark this answers is not in the document]
Comment (novaknole) [reply packages/contracts/contracts/utils/AppStorage.sol:8]: Currently, yes, we have setter and getter for DAO address. but still prefer to keep this as we might need some other stuff to keep it there.

File: packages/contracts/contracts/utils/AppStorage.sol (hunk 1 above)
Comment (novaknole) [reply packages/contracts/contracts/utils/AppStorage.sol:12]: DONE

File: packages/contracts/contracts/utils/UUPSProxy.sol (hunk 3 above)
Comment (novaknole) [reply packages/contracts/contracts/utils/UUPSProxy.sol:9]: done. we used to do it previously and I kept doing it again.

File: packages/contracts/contracts/utils/AppStorage.sol (hunk 2 above)
Comment (novaknole) [reply packages/contracts/contracts/utils/AppStorage.sol:22]: done

File: packages/contracts/contracts/utils/UUPSProxy.sol (hunk 4 above)
Comment (novaknole) [reply packages/contracts/contracts/utils/UUPSProxy.sol:11]: DONE.

File: packages/contracts/contracts/core/plugin/AragonPlugin.sol [hunk 9]
```diff
@@ -0,0 +1,30 @@
+pragma solidity 0.8.10;
+
+import "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
```
Comment (Michael-A-Heuer) [inline packages/contracts/contracts/core/plugin/AragonPlugin.sol:3]: Shouldn't this be `import "@openzeppelin/contracts/proxy/utils/Initializable.sol";` instead?

File: packages/contracts/contracts/core/plugin/AragonPlugin.sol [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+import "@openzeppelin/contracts/utils/Context.sol";
+
+import "../IDAO.sol";
+import "../permission/PermissionManager.sol";
+
+import "../../utils/AppStorage.sol";
+
+/// @title AragonPlugin
+/// @notice An Abtract Aragon Plugin(NON-UPGRADABLE) that plugin developers have to inherit from.
+abstract contract AragonPlugin is AppStorage, Context {
```
Comment (Michael-A-Heuer) [inline packages/contracts/contracts/core/plugin/AragonPlugin.sol:13]: Why are we importing `Initalizable` and not use it here?

File: packages/contracts/contracts/core/plugin/AragonUpgradablePlugin.sol [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+import "../IDAO.sol";
+import "../permission/PermissionManager.sol";
+
+import "../../utils/AppStorage.sol";
+
+/// @title AragonUpgradablePlugin
+/// @notice An Abtract Aragon Plugin(UUPS-UPGRADABLE) that plugin developers have to inherit from.
+abstract contract AragonUpgradablePlugin is Initializable, AppStorage, ContextUpgradeable, UUPSUpgradeable {
+    bytes32 public constant UPGRADE_PERMISSION_ID = keccak256("UPGRADE_PERMISSION");
+
```
Comment (brickpop) [inline packages/contracts/contracts/core/plugin/AragonUpgradablePlugin.sol:16]:
```suggestion
    bytes32 public constant UPGRADE_PERMISSION_ID = keccak256("UPGRADE_PERMISSION");

    // NOTE: Newly added contract variables should always have the size of _gap adapted below as well

```

File: packages/contracts/contracts/utils/AppStorage.sol [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+import "@openzeppelin/contracts/utils/StorageSlot.sol";
+import "../core/IDAO.sol";
+
+/// @notice Core storage that stores information at very specific slots.
+contract AppStorage {    
+    
+    /* Hardcoded constants to save gas
+    bytes32 internal constant DAO_POSITION = keccak256("core.storage.dao");
+    */
+    bytes32 internal constant DAO_POSITION = 0xd69e81f6042b963e91c7595979ec7bb19d41b99e5a44a91c85e5cd5861e49998;
```
Comment (Michael-A-Heuer) [inline packages/contracts/contracts/utils/AppStorage.sol:14]:
```suggestion
    /// @notice The position in storage in which the contract address fulfilling the `IDAO` interface is stored
    /// @dev bytes32 internal constant DAO_POSITION = keccak256("core.storage.dao");
    bytes32 internal constant DAO_POSITION = 0xd69e81f6042b963e91c7595979ec7bb19d41b99e5a44a91c85e5cd5861e49998;
```

File: packages/contracts/contracts/utils/AppStorage.sol [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+    
+    /// @notice Gets the dao(DAO.sol contract) address which is set at DAO_POSITION slot.
+    /// @return IDAO address of the DAO
```
Comment (Michael-A-Heuer) [inline packages/contracts/contracts/utils/AppStorage.sol:17]:
```suggestion
    /// @notice Gets the`IDAO` contract being stored in the `DAO_POSITION` storage slot.
    /// @return The `IDAO` contract.
```

File: packages/contracts/contracts/utils/AppStorage.sol [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+    function dao() public view returns (IDAO) {
+        return IDAO(StorageSlot.getAddressSlot(DAO_POSITION).value);
+    }
+
+    /// @notice Sets the dao(DAO.sol contract) address at DAO_POSITION slot.
```
Comment (Michael-A-Heuer) [inline packages/contracts/contracts/utils/AppStorage.sol:22]:
```suggestion
    /// @notice Stores the`IDAO` contract in the `DAO_POSITION` storage slot.
    /// @param _dao The address of the `IDAO` contract to be stored.
```

File: packages/contracts/contracts/utils/PluginERC1967Proxy.sol [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+pragma solidity 0.8.10;
+
+import "@openzeppelin/contracts/proxy/ERC1967/ERC1967Proxy.sol";
+import "./AppStorage.sol";
+
+/// @title PluginERC1967Proxy
+/// @notice The proxy that delegates calls to the implementation address.
+/// @dev This proxy doesn't include the upgradability logic and presumes
+/// that implementation contracts are UUPSUpgradable... If you want the proxy
+/// to contain upgrade logic, you have to use TransparentUpgradableProxy.
```
Comment (Michael-A-Heuer) [inline packages/contracts/contracts/utils/PluginERC1967Proxy.sol:12]:
```suggestion
/// @notice This contract provides an [ERC-1967](https://eips.ethereum.org/EIPS/eip-1967) compatible proxy contract delegating calls to the implementation address.
/// @dev This proxy doesn't include the upgradability logic and assumes that implementation contracts are `UUPSUpgradable`. 
/// If you want the proxy to contain upgrade logic, you have to use `TransparentUpgradableProxy`.
```

File: packages/contracts/contracts/utils/PluginERC1967Proxy.sol [hunk 16]
```diff
@@ -0,0 +1,25 @@
...
+contract PluginERC1967Proxy is ERC1967Proxy, AppStorage {
+
+    constructor(
```
Comment (Michael-A-Heuer) [inline packages/contracts/contracts/utils/PluginERC1967Proxy.sol:14]:
```suggestion
    /// @notice The constructor initializing the [ERC-1967](https://eips.ethereum.org/EIPS/eip-1967) proxy and setting the `IDAO` address if specified.
    /// @param _dao The DAO address to be stored.
    /// @param _logic The address of the logic contract containing the proxy is pointing to.
    /// @ param _data The data being passed in a delegatecall to `_logic` to initialize the storage (see [OZs `ERC1967Proxy.sol` docs](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/proxy/ERC1967/ERC1967Proxy.sol))
    constructor(
```

File: packages/contracts/deploy/00_managing-dao.ts [hunk 17]
```diff
@@ -17,7 +17,7 @@ const func: DeployFunction = async function (hre: HardhatRuntimeEnvironment) {
     log: true,
     proxy: {
       owner: deployer,
-      proxyContract: 'UUPSProxy',
+      proxyContract: 'PluginERC1967Proxy',
```
Comment (Rekard0) [inline packages/contracts/deploy/00_managing-dao.ts:20]: is it possible this change is not included in this PR?
it is already done in another PR pending review at :
https://github.com/aragon/core/pull/71

same for changes on : 05_ens.ts, 10_dao-registry.ts, 05_aragon-plugin-registry.ts

File: packages/contracts/contracts/core/plugin/AragonUpgradablePlugin.sol [hunk 18]
```diff
@@ -0,0 +1,42 @@
...
+        _;
+    }
+
+    /// @dev Used to check the permissions within the upgradability pattern implementation of OZ
+    function _authorizeUpgrade(address) internal virtual override auth(UPGRADE_PERMISSION_ID) { }
+
+    /// @notice Used by AragonUpgradablePlugin to reserve storage space in 
+    /// case of state variable additions for this contract.
+    /// @dev Note that after adding one or multiple state variables, 
+    /// _gap size below + all state variables this contract uses should be 50.
```
Comment (Michael-A-Heuer) [inline packages/contracts/contracts/core/plugin/AragonUpgradablePlugin.sol:42]:
```suggestion
    /// @notice Used by `AragonUpgradablePlugin` to reserve storage space in case of state variable additions for this contract.
    /// @dev After the addition of state variables, the number of storage slots including `_gap` size must add up to 50.
```

File: packages/contracts/deploy/00_managing-dao.ts (hunk 17 above)
Comment (novaknole) [reply packages/contracts/deploy/00_managing-dao.ts]: DONE

**Review by brickpop — DISMISSED** [review]

File: packages/contracts/contracts/core/plugin/AragonPlugin.sol [hunk 19]
```diff
@@ -0,0 +1,32 @@
...
+import "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+import "@openzeppelin/contracts/utils/Context.sol";
+
+import "../IDAO.sol";
+import "../permission/PermissionManager.sol";
+
+import "../../utils/AppStorage.sol";
+
+/// @title AragonPlugin
+/// @notice An Abtract Aragon Plugin(NON-UPGRADABLE) that plugin developers have to inherit from.
```
Comment (Michael-A-Heuer) [inline packages/contracts/contracts/core/plugin/AragonPlugin.sol:14]:
```suggestion
/// @notice An abstract contract to inherit from when creating a non-upgradable plugin.
```

**Review by Rekard0 — APPROVED** [review]
LGTM!

File: packages/contracts/contracts/core/plugin/AragonUpgradablePlugin.sol [hunk 20]
```diff
@@ -0,0 +1,42 @@
...
+
+import "../../utils/AppStorage.sol";
+import "../IDAO.sol";
+
+/// @title AragonUpgradablePlugin
+/// @notice An Abtract Aragon Plugin(UUPS-UPGRADABLE) that plugin developers have to inherit from.
+abstract contract AragonUpgradablePlugin is Initializable, AppStorage, ContextUpgradeable, UUPSUpgradeable {
+    bytes32 public constant UPGRADE_PERMISSION_ID = keccak256("UPGRADE_PERMISSION");
+
+    // NOTE: Newly added contract variables should always have the size of _gap adapted below as well
```
Comment (Michael-A-Heuer) [inline packages/contracts/contracts/core/plugin/AragonUpgradablePlugin.sol:19]:
```suggestion
    // NOTE: When adding new state variables to the contract, the size of `_gap` has to be adapted below as well.
```

File: packages/contracts/contracts/core/plugin/AragonUpgradablePlugin.sol [hunk 21]
```diff
@@ -0,0 +1,42 @@
...
+import "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+import "@openzeppelin/contracts-upgradeable/utils/ContextUpgradeable.sol";
+
+import { PermissionManager } from "../permission/PermissionManager.sol";
+
+import "../../utils/AppStorage.sol";
+import "../IDAO.sol";
+
+/// @title AragonUpgradablePlugin
+/// @notice An Abtract Aragon Plugin(UUPS-UPGRADABLE) that plugin developers have to inherit from.
```
Comment (Michael-A-Heuer) [inline packages/contracts/contracts/core/plugin/AragonUpgradablePlugin.sol:15]:
```suggestion
/// @notice An abstract contract to inherit from when creating a `UUPSUpgradable` plugin.
```

**Review by Michael-A-Heuer — APPROVED** [review]

**Review by Rekard0 — APPROVED** [review]
